### DISCUSSION TITLE
Is it possible to do "sharp bilinear" upscale?

### DISCUSSION THREAD

**Discussion by realnc** [ImageMagick/ImageMagick#5274] [Help] [answered]
Hello.

I'm trying to scale up images using the "sharp bilinear" method. What that does is first scale the image to the nearest integer multiple of the target size, and then if the result is not equal to the target size, scale using bilinear.

For example, if the input is 320x200 and I want to scale to 1920x1440, it should first scale each dimension separately *without interpolation*. The image should be scaled horizontally to 1920 (320\*6) and vertically to 1400 (200\*7). The result is 1920x1400. This doesn't match the target size, so now it should resize to 1920x1440 with interpolation.

Is this doable? I'm trying to upscale a lot of low resolution images, each having different resolutions, with as little blur as possible.  This is an example when resizing a 256x239 image :

Normal resize is blurry:

![](https://camo.githubusercontent.com/5b2434062e9249bdbe9c37e9819bedd93afc6d3b0e98a9034e53493096074480/68747470733a2f2f696d6167652e6962622e636f2f6a687a44776b2f776974686f75745f7368616465722e706e67)

Sharp bilinear is sharp and without distorted pixels:

![](https://camo.githubusercontent.com/2b18df4bb04dcf9852b3da4bfc7f27c2b093f81513e148a6ca8d5967e7fd1526/68747470733a2f2f696d6167652e6962622e636f2f6855346b39352f776974685f7368616465722e706e67)

**Comment by realnc** [comment, marked as answer]
Oh, wait, I feel so dumb now. Is this what `-scale` does to begin with and all this time I've been trying to do this with `-resize`?

**Comment by snibgo** [comment]
What version of IM, on what platform?

IM has many methods for enlarging an image. You can combine them, of course.

I don't understand "it should first scale each dimension separately *without interpolation*". Perhaps that means you want to duplicate pixels. "-scale" does that.

How did you make your second image? What software, with what settings?

Please show your input image.

**Comment by realnc** [comment]
> What version of IM, on what platform?

7.1.0.20 on Linux.

> I don't understand "it should first scale each dimension separately without interpolation". Perhaps that means you want to duplicate pixels. "-scale" does that.

Yes, duplicate pixels but without the distortion of `-filter Point -resize` when resizing the image to a resolution that is not an integer multiple of the input resolution.  Apparently, `-scale` does that, which for some reason I didn't know before.

Here's what I was doing before. Given an input image with a resolution of 256x239:

![original](https://user-images.githubusercontent.com/1665903/175776079-fc8fb014-caab-4920-a813-739280447832.png)

I want to upscale that to a 4:3 resolution of 1920x1440. Previously, I would do this in two steps. Step 1:

```text
convert original.png -filter Point -resize 1792x1434! temp.png
```

1792 is 256\*7 and 1434 is 239\*6. These are the largest integer scale factors that don't go over the target resolution of 1920x1440. `temp.png` needs to be then be resized again, this time with interpolation so that there's no distorted pixels:

```text
convert tmp.png -resize 1920x1440! sharp_bilinear.png
```

Which results in a nice and sharp image with correct aspect ratio and no pixel distortion:

![sharp_bilinear](https://user-images.githubusercontent.com/1665903/175776230-48cd2a92-9d87-46b7-abef-b814b74e6402.png)

I was looking for a way to somehow automate this without having to bring out the calculator for each image I want to upscale. But as it turns out, I can just do:

```text
convert original.png -scale 1920x1440! scaled.png
```

Which apparently gives the same result:

![scaled](https://user-images.githubusercontent.com/1665903/175776264-2eacbff3-8512-4007-9f1a-6b41b7d4c78a.png)

The "sharp bilinear" term comes from <https://github.com/rsn8887/Sharp-Bilinear-Shaders/blob/master/README.md> and I was trying to replicate this in ImageMagick so I can get sharp upscaled images with non-integer scale factors.

Edit: Well, *almost* the same result. If you zoom in, there are differences. But overall, it serves the same purpose.

**Comment by snibgo** [comment]
> Which apparently gives the same result:
Edit: Well, almost the same result. If you zoom in, there are differences. But overall, it serves the same purpose.

We can compare your two results. I saved your results as m1.png and m2.png, then:
```
magick compare -metric RMSE m1.png m2.png NULL:

229.621 (0.00350379)
```
The RMSE, on a scale of 0 to 1, is 0.0035, so the images are very similar.

Have you tried "-sample 1920x1440!"? You might prefer that result.

**Reply by realnc** [reply]
> Have you tried "-sample 1920x1440!"? You might prefer that result.

This produces wrong results. Pixels are distrorted. It's pretty much exactly what I was trying to avoid :slightly_smiling_face: 

For example, given this input image:

![dos-original](https://user-images.githubusercontent.com/1665903/175777428-c2a4600a-6c1a-4a21-adba-51b76a0bb0f1.png)

If I `-sample`, the dot pattern at the bottom of the screen is all wrong. `-scale` (or the manual "sharp bilinear" method) gets it correct.

**Comment by fmw42** [comment]
If you want to scale up such game images by factor of either 2x or 3x, see -magnify.

**Reply by realnc** [reply]
> If you want to scale up such game images by factor of either 2x or 3x, see -magnify.

Thanks. Will keep it in mind. Usually I need to do aspect ratio correction though and use non-integer scale factors.
